app/upgrades/v18/upgrades.go

```diff
@@ -0,0 +1,116 @@
+		}
+
+		outerMinDelta := hostZone.RedemptionRate.Mul(RedemptionRateOuterMinAdjustment)
+		outerMaxDelta := hostZone.RedemptionRate.Mul(outerAdjustment)
+
+		outerMin := hostZone.RedemptionRate.Sub(outerMinDelta)
+		outerMax := hostZone.RedemptionRate.Add(outerMaxDelta)
+
+		hostZone.MinRedemptionRate = outerMin
+		hostZone.MaxRedemptionRate = outerMax
+
+		k.SetHostZone(ctx, hostZone)
+	}
+}
+
+// Modify HostZoneUnbonding and UserRedemptionRecords NativeTokenAmount to reflect new data structs
+func UpdateUnbondingRecords(
+	ctx sdk.Context,
+	sk stakeibckeeper.Keeper,
+	rk recordskeeper.Keeper,
+	redemptionRatesBeforeProp map[string]map[uint64]sdk.Dec,
+	redemptionRatesDuringProp map[string]sdk.Dec,
+) error {
+	// loop over host zone unbonding records
+	for _, epochUnbondingRecord := range rk.GetAllEpochUnbondingRecord(ctx) {
+		for _, hostZoneUnbonding := range epochUnbondingRecord.HostZoneUnbondings {
+
+			// we can ignore any record that's not currently unbonding
+			if hostZoneUnbonding.Status != recordtypes.HostZoneUnbonding_EXIT_TRANSFER_QUEUE {
+				continue
+			}
+
+			// Grab the redemption rates from before the prop was posted, for a given chain
+			// across all the epochs that unbonded
+			hostZoneRRBeforeProp, ok := redemptionRatesBeforeProp[hostZoneUnbonding.HostZoneId]
+			if !ok {
+				ctx.Logger().Error("Host zone from unbonding record not included in redemption rate mapping")
+				continue
+			}
+
+			// Grab the redemption rate for this specific epoch
+			// If it's not found, that means the unbonding for this epoch occurred after the prop was live
+			recordRedemptionRate, recordUnbondedBeforeProp := hostZoneRRBeforeProp[epochUnbondingRecord.EpochNumber]
+
+			// If we don't have the redemption rate, estimate it
+			if !recordUnbondedBeforeProp {
+				hostZone, found := sk.GetHostZone(ctx, hostZoneUnbonding.HostZoneId)
+				if !found {
+					return errorsmod.Wrapf(stakeibctypes.ErrHostZoneNotFound, "unable to find host zone with chain-id %s", hostZoneUnbonding.HostZoneId)
+				}
+
+				redemptionRateDuringProp := redemptionRatesDuringProp[hostZoneUnbonding.HostZoneId]
+				redemptionRateDuringUpgrade := hostZone.RedemptionRate
+				recordRedemptionRate = redemptionRateDuringProp.Add(redemptionRateDuringUpgrade).Quo(sdk.NewDec(2))
+			}
+
+			// now update all userRedemptionRecords by using the redemption rate to set the native token amount
+			for _, userRedemptionRecordId := range hostZoneUnbonding.UserRedemptionRecords {
+				userRedemptionRecord, found := rk.GetUserRedemptionRecord(ctx, userRedemptionRecordId)
+				if !found {
+					return errorsmod.Wrapf(recordtypes.ErrHostUnbondingRecordNotFound, "unable to find user redemption record with id %s", userRedemptionRecordId)
+				}
+				userRedemptionRecord.NativeTokenAmount = sdk.NewDecFromInt(userRedemptionRecord.StTokenAmount).Mul(recordRedemptionRate).TruncateInt()
+				rk.SetUserRedemptionRecord(ctx, userRedemptionRecord)
+			}
+
```

Comment:
should we also update the `NativeTokenAmount` on the host zone struct? I don't think it's strictly necessary, although feels a bit easier to reason about all of the records lining up!

Reply:
omg, good catch - not sure how I missed that!

Reply:
just pushed up the fix. I also think there was  bug where there was a premature return after calling SetHZUnbonding